frontend/kesaseteli/employer/src/components/application/form/TextInput.tsx

```diff
@@ -1,47 +1,92 @@
+import useApplicationFormField from 'kesaseteli/employer/hooks/application/useApplicationFormField';
+import { useTranslation } from 'next-i18next';
 import React from 'react';
-import { FieldError, RegisterOptions } from 'react-hook-form';
+import {
+  RegisterOptions,
+  useFormContext,
+  UseFormRegister} from 'react-hook-form';
+import { $GridCell,GridCellProps } from 'shared/components/forms/section/FormSection.sc';
 import Application from 'shared/types/employer-application';
 
-import { $TextInput } from './TextInput.sc';
+import { $TextInput, $TextInputProps } from './TextInput.sc';
 
-type InputProps = {
-  validation: RegisterOptions<Application>;
-  id: keyof Application;
-};
 
-const TextInput = ({
+const getComponentType = (type: TextInputProps['type']): typeof HdsTextInput | typeof HdsNumberInput |  typeof HdsTextArea => {
+  switch(type) {
+    case 'number':
+    case 'decimal':
+      return HdsNumberInput;
+
+    case 'textArea':
+      return HdsTextArea;
+
+    case 'text':
+    default:
+      return HdsTextInput;
+  }
+}
+
+export type TextInputProps = {
+  validation?: RegisterOptions<Application>;
+  id: NonNullable<Parameters<UseFormRegister<Application>>[0]>;
+  type?: $TextInputProps['$type'],
+  helperFormat?: string,
+} & GridCellProps;
+
+const TextInput : React.FC<TextInputProps> = ({
   id,
-  validation,
-  ...rest
-}: InputProps & TextInputProps): ReturnType<typeof HdsTextInput> => {
+  validation = {},
+  type = 'text',
+  helperFormat,
+  ...$gridCellProps
+}) => {
+  const { t } = useTranslation();
   const {
-    translateLabel,
-    translateError,
     register,
-    formState: { errors },
-  } = useApplicationForm();
-  const { application, isLoading } = useApplicationApi();
+  } = useFormContext<Application>();
+
+  const { getValue, getError, fieldName } = useApplicationFormField<string>(id);
+  const errorType = getError()?.type;
+
+  const errorText = React.useMemo((): string | undefined => {
+    if (!errorType) {
+      return undefined;
+    }
+    const error = t(`common:application.form.errors.${errorType}`);
+    const helperText = helperFormat ? `${t('common:application.form.helpers.format')}: ${helperFormat}` : undefined;
+    if (['pattern','required'].includes(errorType) && helperText) {
+      return `${error}. ${helperText}`;
+    }
+    return error;
+  },[t,errorType, helperFormat]);
+
+  // TODO: This can be removed after backend supports invalid values in draft save
+  const setValueForBackend = React.useCallback((newValue: string) =>
+    // getError does not always update: https://github.com/react-hook-form/react-hook-form/issues/2893
+    // if value hasnt changed (getValue is same as new value), then error is present and invalid value is changed to undefined
+    // to prevent backend to fail
+    getError() && getValue() === newValue ? undefined : newValue,
+   [getError,getValue]);
```

Comment:
You may run this file through Prettier.

Reply:
prettier should be run on pre-commit hook and most of the times it works but somehow it fails sometimes. Not sure if there is something wrong with pre-hook or is it only on my local machine.